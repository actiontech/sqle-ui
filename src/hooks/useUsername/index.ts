import { useBoolean } from 'ahooks';
import React from 'react';
import { IUserTipResV1 } from '../../api/common';
import user from '../../api/user';
import { ResponseCode } from '../../data/common';

const useUsername = () => {
  const [usernameList, setUsernameList] = React.useState<IUserTipResV1[]>([]);
  const [loading, { setTrue, setFalse }] = useBoolean();

  const updateUsernameList = React.useCallback(() => {
    setTrue();
    user
      .getUserTipListV1()
      .then((res) => {
        if (res.data.code === ResponseCode.SUCCESS) {
          setUsernameList(res.data?.data ?? []);
        } else {
          setUsernameList([]);
        }
      })
      .catch(() => {
        setUsernameList([]);
      })
      .finally(() => {
        setFalse();
      });
  }, [setFalse, setTrue]);

  return {
    usernameList,
    loading,
    updateUsernameList,
  };
};

export default useUsername;
