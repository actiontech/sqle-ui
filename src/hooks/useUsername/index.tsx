import { useBoolean } from 'ahooks';
import { Select } from 'antd';
import React from 'react';
import { IUserTipResV1 } from '../../api/common';
import user from '../../api/user';
import { ResponseCode } from '../../data/common';

const useUsername = () => {
  const [usernameList, setUsernameList] = React.useState<IUserTipResV1[]>([]);
  const [loading, { setTrue, setFalse }] = useBoolean();

  const updateUsernameList = React.useCallback(() => {
    setTrue();
    user
      .getUserTipListV1({})
      .then((res) => {
        if (res.data.code === ResponseCode.SUCCESS) {
          setUsernameList(res.data?.data ?? []);
        } else {
          setUsernameList([]);
        }
      })
      .catch(() => {
        setUsernameList([]);
      })
      .finally(() => {
        setFalse();
      });
  }, [setFalse, setTrue]);

  const generateUsernameSelectOption = React.useCallback(() => {
    return usernameList.map((user) => {
      return (
        <Select.Option key={user.user_name} value={user.user_name ?? ''}>
          {user.user_name}
        </Select.Option>
      );
    });
  }, [usernameList]);

  return {
    usernameList,
    loading,
    updateUsernameList,
    generateUsernameSelectOption,
  };
};

export default useUsername;
