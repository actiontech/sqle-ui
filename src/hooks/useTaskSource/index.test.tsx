import { act, renderHook } from '@testing-library/react-hooks';
import {
  render,
  fireEvent,
  screen,
  act as reactAct,
  cleanup,
} from '@testing-library/react';
import {
  mockUseTaskSource,
  rejectThreeSecond,
  resolveErrorThreeSecond,
} from '../../testUtils/mockRequest';
import { Select } from 'antd';
import useUsername from '.';

describe('useUserGroup', () => {
  let requestSpy: jest.SpyInstance;
  beforeEach(() => {
    jest.useFakeTimers();
    requestSpy = mockUseTaskSource();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('should get task source from request', async () => {
    const { result, waitForNextUpdate } = renderHook(() => useUsername());
    expect(result.current.loading).toBe(false);
    expect(result.current.taskSourceList).toEqual([]);
    const { baseElement } = render(
      <Select>{result.current.generateTaskSourceSelectOption()}</Select>
    );
    expect(baseElement).toMatchSnapshot();

    act(() => {
      result.current.updateTaskSourceList();
    });

    expect(result.current.loading).toBe(true);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.taskSourceList).toEqual([]);

    jest.advanceTimersByTime(3000);
    await waitForNextUpdate();

    expect(result.current.loading).toBe(false);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.taskSourceList).toEqual([{ source: 'source1' }]);
    cleanup();

    const { baseElement: baseElementWithOptions } = render(
      <Select data-testid="testId" value="value1">
        {result.current.generateTaskSourceSelectOption()}
      </Select>
    );
    expect(baseElementWithOptions).toMatchSnapshot();

    reactAct(() => {
      fireEvent.mouseDown(screen.getByText('value1'));
      jest.runAllTimers();
    });

    await screen.findAllByText('source1');
    expect(baseElementWithOptions).toMatchSnapshot();
  });

  test('should set list to empty array when response code is not equal success code', async () => {
    const { result, waitForNextUpdate } = renderHook(() => useUsername());
    expect(result.current.loading).toBe(false);
    expect(result.current.taskSourceList).toEqual([]);

    act(() => {
      result.current.updateTaskSourceList();
    });

    expect(result.current.loading).toBe(true);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.taskSourceList).toEqual([]);

    jest.advanceTimersByTime(3000);
    await waitForNextUpdate();

    expect(result.current.loading).toBe(false);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.taskSourceList).toEqual([{ source: 'source1' }]);
    requestSpy.mockClear();
    requestSpy.mockImplementation(() =>
      resolveErrorThreeSecond([{ source: 'source1' }])
    );

    act(() => {
      result.current.updateTaskSourceList();
    });
    expect(result.current.loading).toBe(true);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.taskSourceList).toEqual([
      {
        source: 'source1',
      },
    ]);

    jest.advanceTimersByTime(3000);
    await waitForNextUpdate();

    expect(result.current.loading).toBe(false);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.taskSourceList).toEqual([]);
  });

  test('should set list to empty array when response throw error', async () => {
    const { result, waitForNextUpdate } = renderHook(() => useUsername());
    expect(result.current.loading).toBe(false);
    expect(result.current.taskSourceList).toEqual([]);

    act(() => {
      result.current.updateTaskSourceList();
    });

    expect(result.current.loading).toBe(true);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.taskSourceList).toEqual([]);

    jest.advanceTimersByTime(3000);
    await waitForNextUpdate();

    expect(result.current.loading).toBe(false);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.taskSourceList).toEqual([{ source: 'source1' }]);
    requestSpy.mockClear();
    requestSpy.mockImplementation(() =>
      rejectThreeSecond([{ source: 'source1' }])
    );

    act(() => {
      result.current.updateTaskSourceList();
    });
    expect(result.current.loading).toBe(true);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.taskSourceList).toEqual([
      {
        source: 'source1',
      },
    ]);

    jest.advanceTimersByTime(3000);
    await waitForNextUpdate();

    expect(result.current.loading).toBe(false);
    expect(requestSpy).toBeCalledTimes(1);
    expect(result.current.taskSourceList).toEqual([]);
  });
});
