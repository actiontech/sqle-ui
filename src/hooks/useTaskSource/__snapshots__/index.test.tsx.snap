// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTaskSource should generate dbTypes select options with source 1`] = `
<body>
  <div>
    <div
      class="ant-select ant-select-single ant-select-show-arrow ant-select-open"
    >
      <div
        class="ant-select-selector"
      >
        <span
          class="ant-select-selection-search"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-expanded="true"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="ant-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="ant-select-selection-item"
          title="type1"
        >
          type1
        </span>
      </div>
      <span
        aria-hidden="true"
        class="ant-select-arrow"
        style="user-select: none;"
        unselectable="on"
      >
        <span
          aria-label="down"
          class="anticon anticon-down ant-select-suffix"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="down"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
            />
          </svg>
        </span>
      </span>
    </div>
  </div>
  <div
    style="position: absolute; top: 0px; left: 0px; width: 100%;"
  >
    <div>
      <div
        class="ant-select-dropdown"
        style="opacity: 0; min-width: 0; width: 0px;"
      >
        <div>
          <div
            id="rc_select_TEST_OR_SSR_list"
            role="listbox"
            style="height: 0px; width: 0px; overflow: hidden;"
          >
            <div
              aria-selected="false"
              id="rc_select_TEST_OR_SSR_list_0"
              role="option"
            >
              mysql
            </div>
          </div>
          <div
            class="rc-virtual-list"
            style="position: relative;"
          >
            <div
              class="rc-virtual-list-holder"
              style="max-height: 256px; overflow-y: hidden;"
            >
              <div>
                <div
                  class="rc-virtual-list-holder-inner"
                  style="display: flex; flex-direction: column;"
                >
                  <div
                    aria-selected="false"
                    class="ant-select-item ant-select-item-option ant-select-item-option-active"
                  >
                    <div
                      class="ant-select-item-option-content"
                    >
                      <div
                        class="database-type-logo-wrapper"
                      >
                        <img
                          alt=""
                          src="/v1/static/instance_logo?instance_type=mysql"
                        />
                        <span
                          class="ant-select-item-option-content"
                          title="mysql"
                        >
                          mysql
                        </span>
                      </div>
                    </div>
                    <span
                      aria-hidden="true"
                      class="ant-select-item-option-state"
                      style="user-select: none;"
                      unselectable="on"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="rc-virtual-list-scrollbar"
              style="width: 8px; top: 0px; bottom: 0px; right: 0px; position: absolute; display: none;"
            >
              <div
                class="rc-virtual-list-scrollbar-thumb"
                style="width: 100%; height: 128px; top: 0px; left: 0px; position: absolute; background: rgba(0, 0, 0, 0.5); border-radius: 99px; cursor: pointer; user-select: none;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`useTaskSource should generate dbTypes select options with source 2`] = `
<body>
  <div>
    <div
      class="ant-select ant-select-single ant-select-show-arrow ant-select-open"
    >
      <div
        class="ant-select-selector"
      >
        <span
          class="ant-select-selection-search"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-expanded="true"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="ant-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="ant-select-selection-item"
          title="type3"
        >
          type3
        </span>
      </div>
      <span
        aria-hidden="true"
        class="ant-select-arrow"
        style="user-select: none;"
        unselectable="on"
      >
        <span
          aria-label="down"
          class="anticon anticon-down ant-select-suffix"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="down"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
            />
          </svg>
        </span>
      </span>
    </div>
  </div>
  <div
    style="position: absolute; top: 0px; left: 0px; width: 100%;"
  >
    <div>
      <div
        class="ant-select-dropdown"
        style="opacity: 0; min-width: 0; width: 0px;"
      >
        <div>
          <div
            id="rc_select_TEST_OR_SSR_list"
            role="listbox"
            style="height: 0px; width: 0px; overflow: hidden;"
          >
            <div
              aria-selected="false"
              id="rc_select_TEST_OR_SSR_list_0"
              role="option"
            >
              oracle
            </div>
          </div>
          <div
            class="rc-virtual-list"
            style="position: relative;"
          >
            <div
              class="rc-virtual-list-holder"
              style="max-height: 256px; overflow-y: hidden;"
            >
              <div>
                <div
                  class="rc-virtual-list-holder-inner"
                  style="display: flex; flex-direction: column;"
                >
                  <div
                    aria-selected="false"
                    class="ant-select-item ant-select-item-option ant-select-item-option-active"
                  >
                    <div
                      class="ant-select-item-option-content"
                    >
                      <div
                        class="database-type-logo-wrapper"
                      >
                        <img
                          alt=""
                          src="/v1/static/instance_logo?instance_type=oracle"
                        />
                        <span
                          class="ant-select-item-option-content"
                          title="oracle"
                        >
                          oracle
                        </span>
                      </div>
                    </div>
                    <span
                      aria-hidden="true"
                      class="ant-select-item-option-state"
                      style="user-select: none;"
                      unselectable="on"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="rc-virtual-list-scrollbar"
              style="width: 8px; top: 0px; bottom: 0px; right: 0px; position: absolute; display: none;"
            >
              <div
                class="rc-virtual-list-scrollbar-thumb"
                style="width: 100%; height: 128px; top: 0px; left: 0px; position: absolute; background: rgba(0, 0, 0, 0.5); border-radius: 99px; cursor: pointer; user-select: none;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`useTaskSource should generate dbTypes select options with source 3`] = `
<body>
  <div>
    <div
      class="ant-select ant-select-single ant-select-show-arrow ant-select-open"
    >
      <div
        class="ant-select-selector"
      >
        <span
          class="ant-select-selection-search"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_-1"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-expanded="true"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="ant-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="ant-select-selection-item"
          title="type3"
        >
          type3
        </span>
      </div>
      <span
        aria-hidden="true"
        class="ant-select-arrow"
        style="user-select: none;"
        unselectable="on"
      >
        <span
          aria-label="down"
          class="anticon anticon-down ant-select-suffix"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="down"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
            />
          </svg>
        </span>
      </span>
    </div>
  </div>
  <div
    style="position: absolute; top: 0px; left: 0px; width: 100%;"
  >
    <div>
      <div
        class="ant-select-dropdown ant-select-dropdown-empty"
        style="opacity: 0; min-width: 0; width: 0px;"
      >
        <div>
          <div
            class="ant-select-item-empty"
            id="rc_select_TEST_OR_SSR_list"
            role="listbox"
            style=""
          >
            <div
              class="ant-empty ant-empty-normal ant-empty-small"
            >
              <div
                class="ant-empty-image"
              >
                <svg
                  class="ant-empty-img-simple"
                  height="41"
                  viewBox="0 0 64 41"
                  width="64"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g
                    fill="none"
                    fill-rule="evenodd"
                    transform="translate(0 1)"
                  >
                    <ellipse
                      class="ant-empty-img-simple-ellipse"
                      cx="32"
                      cy="33"
                      rx="32"
                      ry="7"
                    />
                    <g
                      class="ant-empty-img-simple-g"
                      fill-rule="nonzero"
                    >
                      <path
                        d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                      />
                      <path
                        class="ant-empty-img-simple-path"
                        d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                      />
                    </g>
                  </g>
                </svg>
              </div>
              <div
                class="ant-empty-description"
              >
                No data
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`useTaskSource should get task source from request 1`] = `
<body>
  <div>
    <div
      class="ant-select ant-select-single ant-select-show-arrow"
    >
      <div
        class="ant-select-selector"
      >
        <span
          class="ant-select-selection-search"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="ant-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="ant-select-selection-placeholder"
        />
      </div>
      <span
        aria-hidden="true"
        class="ant-select-arrow"
        style="user-select: none;"
        unselectable="on"
      >
        <span
          aria-label="down"
          class="anticon anticon-down ant-select-suffix"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="down"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
            />
          </svg>
        </span>
      </span>
    </div>
  </div>
</body>
`;

exports[`useTaskSource should get task source from request 2`] = `
<body>
  <div>
    <div
      class="ant-select ant-select-single ant-select-show-arrow"
      data-testid="testId"
    >
      <div
        class="ant-select-selector"
      >
        <span
          class="ant-select-selection-search"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="ant-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="ant-select-selection-item"
          title="value1"
        >
          value1
        </span>
      </div>
      <span
        aria-hidden="true"
        class="ant-select-arrow"
        style="user-select: none;"
        unselectable="on"
      >
        <span
          aria-label="down"
          class="anticon anticon-down ant-select-suffix"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="down"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
            />
          </svg>
        </span>
      </span>
    </div>
  </div>
</body>
`;

exports[`useTaskSource should get task source from request 3`] = `
<body>
  <div>
    <div
      class="ant-select ant-select-single ant-select-show-arrow ant-select-open"
      data-testid="testId"
    >
      <div
        class="ant-select-selector"
      >
        <span
          class="ant-select-selection-search"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-expanded="true"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="ant-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            type="search"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="ant-select-selection-item"
          title="value1"
        >
          value1
        </span>
      </div>
      <span
        aria-hidden="true"
        class="ant-select-arrow"
        style="user-select: none;"
        unselectable="on"
      >
        <span
          aria-label="down"
          class="anticon anticon-down ant-select-suffix"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="down"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
            />
          </svg>
        </span>
      </span>
    </div>
  </div>
  <div
    style="position: absolute; top: 0px; left: 0px; width: 100%;"
  >
    <div>
      <div
        class="ant-select-dropdown"
        style="opacity: 0; min-width: 0; width: 0px;"
      >
        <div>
          <div
            id="rc_select_TEST_OR_SSR_list"
            role="listbox"
            style="height: 0px; width: 0px; overflow: hidden;"
          >
            <div
              aria-label="source1"
              aria-selected="false"
              id="rc_select_TEST_OR_SSR_list_0"
              role="option"
            >
              source1
            </div>
            <div
              aria-label="source2"
              aria-selected="false"
              id="rc_select_TEST_OR_SSR_list_1"
              role="option"
            >
              source2
            </div>
          </div>
          <div
            class="rc-virtual-list"
            style="position: relative;"
          >
            <div
              class="rc-virtual-list-holder"
              style="max-height: 256px; overflow-y: hidden;"
            >
              <div>
                <div
                  class="rc-virtual-list-holder-inner"
                  style="display: flex; flex-direction: column;"
                >
                  <div
                    aria-selected="false"
                    class="ant-select-item ant-select-item-option ant-select-item-option-active"
                    title="source1"
                  >
                    <div
                      class="ant-select-item-option-content"
                    >
                      source1
                    </div>
                    <span
                      aria-hidden="true"
                      class="ant-select-item-option-state"
                      style="user-select: none;"
                      unselectable="on"
                    />
                  </div>
                  <div
                    aria-selected="false"
                    class="ant-select-item ant-select-item-option"
                    title="source2"
                  >
                    <div
                      class="ant-select-item-option-content"
                    >
                      source2
                    </div>
                    <span
                      aria-hidden="true"
                      class="ant-select-item-option-state"
                      style="user-select: none;"
                      unselectable="on"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="rc-virtual-list-scrollbar"
              style="width: 8px; top: 0px; bottom: 0px; right: 0px; position: absolute; display: none;"
            >
              <div
                class="rc-virtual-list-scrollbar-thumb"
                style="width: 100%; height: 128px; top: 0px; left: 0px; position: absolute; background: rgba(0, 0, 0, 0.5); border-radius: 99px; cursor: pointer; user-select: none;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
